src/bridge/cockpitcgroupsamples.c

```diff
@@ -21,119 +21,232 @@
-  double ret;
-  gsize len;
-
-  path = g_build_filename (prefix, suffix, NULL);
-  if (!g_file_get_contents (path, &file_contents, &len, &error))
+  int fd = -1;
+  ssize_t len;
+  const char *ret = NULL;
+
+  fd = openat (dirfd, fname, O_RDONLY);
+  if (fd < 0)
     {
-      if (g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_NOENT) ||
-          g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_NODEV))
-        g_debug ("samples file not found: %s", path);
+      if (errno == ENOENT || errno == ENODEV)
+        g_debug ("samples file not found: %s/%s", cgroup, fname);
       else
-        g_message ("error loading file: %s: %s", path, error->message);
-      g_error_free (error);
-      ret = -1;
+        g_message ("error opening file: %s/%s: %m", cgroup, fname);
+      return NULL;
     }
-  else
+
+  /* don't do fancy retry/error handling here -- we know what cgroupfs attributes look like,
+   * it's a virtual file system (does not block/no multiple reads), and it's ok to miss
+   * one sample due to EINTR or some race condition */
+  len = read (fd, buf, bufsize);
+  if (len < 0)
     {
-      ret = g_ascii_strtod (file_contents, NULL);
+      g_message ("error loading file: %s/%s: %m", cgroup, fname);
+      goto out;
     }
+  /* we really expect a much smaller read; if we get a full buffer, there's likely
+   * more data, and we are misinterpreting stuff */
+  if (len >= bufsize)
+    {
+      g_warning ("cgroupfs value %s/%s is too large", cgroup, fname);
+      goto out;
+    }
+  buf[len] = '\0';
+  ret = buf;
 
-  g_free (file_contents);
-  g_free (path);
-
+out:
+  if (fd >= 0)
+    close (fd);
   return ret;
 }
 
-static void
-collect_memory (CockpitSamples *samples,
-                const gchar *path,
-                const gchar *cgroup)
+
+static gint64
+read_int64 (int dirfd,
+            const char *cgroup,
+            const char *fname)
+{
+  char buf[30];
+  const char *contents = read_file (dirfd, buf, sizeof buf, cgroup, fname);
+
+  if (contents == NULL)
+      return -1;
+  /* no error checking; these often have values like "max" which we want to treat as "invalid/absent" */
+  return atoll(contents);
+}
+
+static gint64
+read_keyed_int64 (int dirfd,
+                  const char *cgroup,
+                  const char *fname,
+                  const char *key)
 {
-  double mem_usage_in_bytes;
-  double mem_limit_in_bytes;
-  double memsw_usage_in_bytes;
-  double memsw_limit_in_bytes;
+  char buf[256];
+  const char *contents = read_file (dirfd, buf, sizeof buf, cgroup, fname);
+  const char *offset;
+  size_t key_len = strlen (key);
+  char *endptr = NULL;
+  gint64 result;
+
+  if (contents == NULL)
+      return -1;
+
+  /* search for a word match of key */
+  offset = contents;
+  for (;;) {
+      offset = strstr (offset, key);
+      if (offset == NULL)
+        return -1;
+      /* either matches at start of string, or after a line break */
+      if (offset == contents || offset[-1] == '\n')
```

Comment:
This looks weird because `offset` is a bad name for this variable.  `offset` should be an int, but this is a pointer to a (potentially matching) key.  Might be worth a rename.

Reply:
Good point. Renamed to "match".